desc/objectives/nae_utils.py

```diff
@@ -0,0 +1,239 @@
+    bases : dict
+        dictionary of Rbasis_cos, Rbasis_sin, Zbasis_cos, Zbasis_sin,
+        the FourierSeries basis objects used to obtain the coefficients, where
+        _cos or _sin denotes the symmetry of the (toroidal) Fourier series.
+        symmetry is such that the R or Z coefficients is stellarator symmetric
+        i.e. R_1_1_n uses the Rbasis_cos, since cos(theta)*cos(phi) is
+        stellarator symmetric for R i.e. R(-theta,-phi) = R(theta,phi)
+        and Z_1_1_n uses the Zbasis_sin as the term is cos(theta)*sin(phi)
+        since Z(-theta,-phi) = - Z(theta,phi) for Z stellarator symmetry.
+
+    Returns
+    -------
+    Rconstraints : tuple of Objective
+        tuple of constraints of type FixSumModesR, which enforce
+        the O(rho) behavior of the equilibrium R coefficents to match the NAE.
+    Zconstraints : tuple of Objective
+        tuple of constraints of type FixSumModesZ, which enforce
+        the O(rho) behavior of the equilibrium Z coefficents to match the NAE.
+    """
+    # r is the ratio  r_NAE / rho_DESC
+    r = np.sqrt(2 * desc_eq.Psi / qsc.Bbar / 2 / np.pi)
+
+    Rconstraints = ()
+    Zconstraints = ()
+    Rbasis_cos = bases["Rbasis_cos"]
+    Zbasis_cos = bases["Zbasis_cos"]
+    Rbasis_sin = bases["Rbasis_sin"]
+    Zbasis_sin = bases["Zbasis_sin"]
+
+    # R_11n
+    for n, NAEcoeff in zip(Rbasis_cos.modes[:, 2], coeffs["R_1_1_n"]):
+        sum_weights = []
+        modes = []
+        target = NAEcoeff * r
+        for k in range(1, int((desc_eq.L + 1) / 2) + 1):  # FIXME: add by 1 desc_eq.L+1
+            modes.append([2 * k - 1, 1, n])
+            sum_weights.append([(-1) ** k * k])
+        modes = np.atleast_2d(modes)
+        sum_weights = -np.atleast_1d(sum_weights)
+        Rcon = FixSumModesR(target=target, sum_weights=sum_weights, modes=modes)
+        desc_eq.R_lmn
+        Rconstraints += (Rcon,)
+    # Z_1-1n
+    for n, NAEcoeff in zip(Zbasis_cos.modes[:, 2], coeffs["Z_1_neg1_n"]):
+        sum_weights = []
+        modes = []
+        target = NAEcoeff * r
+        for k in range(1, int((desc_eq.L + 1) / 2) + 1):
+            modes.append([2 * k - 1, -1, n])
+            sum_weights.append([(-1) ** k * k])
+        modes = np.atleast_2d(modes)
+        sum_weights = -np.atleast_1d(sum_weights)
+        Zcon = FixSumModesZ(target=target, sum_weights=sum_weights, modes=modes)
+        Zconstraints += (Zcon,)
+
+    # R_1-1n
+    for n, NAEcoeff in zip(Rbasis_sin.modes[:, 2], coeffs["R_1_neg1_n"]):
+        sum_weights = []
+        modes = []
+        target = NAEcoeff * r
+        for k in range(1, int((desc_eq.L + 1) / 2) + 1):
+            modes.append([2 * k - 1, -1, n])
+            sum_weights.append([(-1) ** k * k])
+        modes = np.atleast_2d(modes)
+        sum_weights = -np.atleast_1d(sum_weights)
+        Rcon = FixSumModesR(target=target, sum_weights=sum_weights, modes=modes)
+        Rconstraints += (Rcon,)
+    # Z_11n
```

Comment:
I think these comments and variable names will be clearer if you use the consistent notation `Z_1_+1_n` etc.

Reply:
I made the comments consistent with the variable names